media: support application/manifest+json

Seirdy opened the pull request:
The standard file extension for Web App Manifest files is
".webmanifest". This commit allows Hugo to recognize .webmanifest files
as "application/manifest+json" files and to minify them using its
JSON minifier.

The .webmanifest file extension is recommended in the w3c spec to
simplify media type registration:
https://www.w3.org/TR/appmanifest/#media-type-registration

Webhint docs are also relevant:
https://webhint.io/docs/user-guide/hints/hint-manifest-file-extension/

Closes #8624

bep commented on output/outputFormat.go:
```diff
@@ -143,6 +143,14 @@ var (
 	}
 
+	WebAppManifestFormat = Format{
+		Name:        "Web App Manifest",
```
I suggest name `WebAppManifest` which should make life easier.

Also, could you add this type to the `DefaultFormats`, also I suspect there will be a test failure or two.

bep replied:
OK, sorry for the confusion. The type should follow the naming standard of the others (that is a discussion for another time :-); I meant the `Name`,  which is now `Web App Manifest`.

Seirdy commented:
On Fri, Jun 11, 2021 at 09:55:16AM -0700, Bjørn Erik Pedersen wrote:
***@***.*** commented on this pull request.
>I suggest name `WebAppManifest` which should make life easier.
>
>Also, could you add this type to the `DefaultFormats`, also I suspect there will be a test failure or two.

Did both; just pushed. Some tests fail on my end; I'll attach the 
results.


-- 
/Seirdy

Seirdy commented:
Test results:

```
?   	github.com/gohugoio/hugo	[no test files]
ok  	github.com/gohugoio/hugo/bufferpool	(cached)
ok  	github.com/gohugoio/hugo/cache/filecache	(cached)
ok  	github.com/gohugoio/hugo/cache/namedmemcache	(cached)
ok  	github.com/gohugoio/hugo/codegen	(cached)
ok  	github.com/gohugoio/hugo/commands	(cached)
ok  	github.com/gohugoio/hugo/common/collections	(cached)
?   	github.com/gohugoio/hugo/common/constants	[no test files]
ok  	github.com/gohugoio/hugo/common/herrors	(cached)
?   	github.com/gohugoio/hugo/common/hexec	[no test files]
ok  	github.com/gohugoio/hugo/common/hreflect	(cached)
?   	github.com/gohugoio/hugo/common/hugio	[no test files]
ok  	github.com/gohugoio/hugo/common/hugo	(cached)
ok  	github.com/gohugoio/hugo/common/loggers	(cached)
ok  	github.com/gohugoio/hugo/common/maps	(cached)
ok  	github.com/gohugoio/hugo/common/math	(cached)
ok  	github.com/gohugoio/hugo/common/para	(cached)
?   	github.com/gohugoio/hugo/common/terminal	[no test files]
ok  	github.com/gohugoio/hugo/common/text	(cached)
ok  	github.com/gohugoio/hugo/common/types	(cached)
?   	github.com/gohugoio/hugo/common/urls	[no test files]
ok  	github.com/gohugoio/hugo/compare	(cached)
ok  	github.com/gohugoio/hugo/config	(cached)
ok  	github.com/gohugoio/hugo/config/privacy	(cached)
ok  	github.com/gohugoio/hugo/config/services	(cached)
ok  	github.com/gohugoio/hugo/create	(cached)
ok  	github.com/gohugoio/hugo/deploy	(cached)
ok  	github.com/gohugoio/hugo/deps	(cached)
?   	github.com/gohugoio/hugo/docshelper	[no test files]
ok  	github.com/gohugoio/hugo/helpers	0.042s
?   	github.com/gohugoio/hugo/htesting	[no test files]
?   	github.com/gohugoio/hugo/htesting/hqt	[no test files]
ok  	github.com/gohugoio/hugo/hugofs	(cached)
ok  	github.com/gohugoio/hugo/hugofs/files	(cached)
ok  	github.com/gohugoio/hugo/hugofs/glob	(cached)


ERROR 2021/06/11 18:28:21 You have configured output formats for 'taxonomy' in your site configuration. In Hugo 0.73.0 we fixed these to be what most people expect (taxonomy and term).
But this also means that your site configuration may not do what you expect. If it is correct, you can suppress this message by following the instructions below.
If you feel that this should not be logged as an ERROR, you can ignore it by adding this to your site config:
ignoreErrors = ["error-output-taxonomy"]




ERROR 2021/06/11 18:28:21 You have the value 'taxonomy' in the disabledKinds list. In Hugo 0.73.0 we fixed these to be what most people expect (taxonomy and term).
But this also means that your site configuration may not do what you expect. If it is correct, you can suppress this message by following the instructions below.
If you feel that this should not be logged as an ERROR, you can ignore it by adding this to your site config:
ignoreErrors = ["error-disable-taxonomy"]


Source changed "content/post/_index.md": WRITE
Source changed "content/post/dir/p1.md": WRITE
Source changed "content/post/_index.md": WRITE
Source changed "content/post/_index.md": WRITE
Source changed "content/post/_index.md": WRITE
Template changed "/mywork/layouts/partials/mypartial1.html": WRITE
Template changed "/mywork/layouts/partials/mypartial2.html": WRITE
Template changed "/mywork/layouts/_default/_markup/render-link.html": WRITE
Template changed "/mywork/layouts/_default/_markup/render-image.html": WRITE
Template changed "/mywork/layouts/docs/_markup/render-link.html": WRITE
Template changed "/mywork/layouts/partials/mypartial1.html": WRITE
Template changed "/mywork/layouts/partials/mypartial3.html": WRITE
Template changed "/mywork/layouts/partials/mypartial4.html": WRITE
Template changed "/mywork/layouts/shortcodes/myshortcode3.html": WRITE
Template changed "/mywork/layouts/_default/_markup/render-link.html": REMOVE
Template added "/mywork/layouts/_default/_markup/render-link.html": WRITE
WARN 2021/06/11 18:28:21 Unsupported symlink found in "realsymmod", skipping.
ERROR 2021/06/11 18:28:21 "layouts/_default/baseof.html:4:1": parse failed: template: _default/single.html:4: unexpected "}" in operand
ERROR 2021/06/11 18:28:21 "layouts/_default/baseof.html:4:1": parse failed: template: _default/single.html:4: unexpected "}" in operand
ERROR 2021/06/11 18:28:22 render of "page" failed: "layouts/_default/baseof.html:4:15": execute of template failed: template: _default/single.html:4:15: executing "_default/single.html" at <.Titles>: can't evaluate field Titles in type *hugolib.pageState
ERROR 2021/06/11 18:28:22 render of "page" failed: "layouts/_default/baseof.html:4:15": execute of template failed: template: _default/single.html:4:15: executing "_default/single.html" at <.Titles>: can't evaluate field Titles in type *hugolib.pageState
ERROR 2021/06/11 18:28:22 render of "page" failed: "layouts/_default/single.html:5:14": execute of template failed: template: _default/single.html:5:14: executing "main" at <.Titles>: can't evaluate field Titles in type *hugolib.pageState
ERROR 2021/06/11 18:28:22 render of "page" failed: "layouts/_default/single.html:5:14": execute of template failed: template: _default/single.html:5:14: executing "main" at <.Titles>: can't evaluate field Titles in type *hugolib.pageState
ERROR 2021/06/11 18:28:22 render of "page" failed: "layouts/_default/single.html:5:14": execute of template failed: template: _default/single.html:5:14: executing "main" at <.ThisIsAVeryLongTitle>: can't evaluate field ThisIsAVeryLongTitle in type *hugolib.pageState
ERROR 2021/06/11 18:28:22 render of "page" failed: "layouts/_default/single.html:5:14": execute of template failed: template: _default/single.html:5:14: executing "main" at <.ThisIsAVeryLongTitle>: can't evaluate field ThisIsAVeryLongTitle in type *hugolib.pageState
ERROR 2021/06/11 18:28:22 render of "page" failed: "layouts/_default/single.html:5:21": execute of template failed: template: _default/single.html:5:21: executing "main" at <.Parent.Parent.Parent>: nil pointer evaluating page.Page.Parent
ERROR 2021/06/11 18:28:22 render of "page" failed: "layouts/_default/single.html:5:21": execute of template failed: template: _default/single.html:5:21: executing "main" at <.Parent.Parent.Parent>: nil pointer evaluating page.Page.Parent
Source changed "content/sect/doc2.en.md": REMOVE
Source changed "content/new1.en.md": CREATE
Source changed "content/new2.en.md": CREATE
Source changed "content/new1.fr.md": CREATE
Source changed "content/sect/doc1.en.md": WRITE
Source changed "content/new1renamed.en.md": RENAME
Source changed "content/new1.en.md": RENAME
Template changed "layouts/_default/single.html": WRITE
i18n changed "i18n/fr.yaml": WRITE
Template changed "layouts/shortcodes/shortcode.html": WRITE
Source changed "content/blog/page1.md": WRITE
Data changed "data/mydata.toml": WRITE
Template changed "layouts/_default/single.html": WRITE
Template changed "layouts/_default/baseof.html": WRITE
Template changed "layouts/partials/mypartial3.html": WRITE
Template changed "layouts/shortcodes/output.amp.html": WRITE
Source changed "content/pages/p3/index.md": WRITE
Source changed "content/pages/p4/index.md": WRITE
Source changed "content/pages/p5.md": WRITE
Source changed "content/pages/p6.md": WRITE
Markup type mmark is deprecated and will be removed in a future release. See https://gohugo.io//content-management/formats/#list-of-content-formats
.File.Filename on zero object. Wrap it in if or with: {{ with .File }}{{ .Filename }}{{ end }}
WARN 2021/06/11 18:28:28 Content directory "content/en/b2" have both index.* and _index.* files, pick one.
WARN 2021/06/11 18:28:28 found no layout file for "HTML" for kind "home": You should create a template file which matches Hugo Layouts Lookup Rules for this combination.
WARN 2021/06/11 18:28:28 found no layout file for "HTML" for kind "taxonomy": You should create a template file which matches Hugo Layouts Lookup Rules for this combination.
WARN 2021/06/11 18:28:28 found no layout file for "HTML" for kind "taxonomy": You should create a template file which matches Hugo Layouts Lookup Rules for this combination.
Markup type mmark is deprecated and will be removed in a future release. See https://gohugo.io//content-management/formats/#list-of-content-formats
--- FAIL: TestShortcodesInSite (0.41s)
    shortcode_test.go:631: 
        error:
          got non-nil error
        got:
          e"logged 19 error(s)"
        stack:
          /home/rkumar/Executables/ghq/github.com/gohugoio/hugo/hugolib/testhelpers_test.go:954
            b.Assert(h.Build(buildCfg), qt.IsNil)
          /home/rkumar/Executables/ghq/github.com/gohugoio/hugo/hugolib/testhelpers_test.go:928
            return buildSingleSiteExpected(t, false, false, depsCfg, buildCfg)
          /home/rkumar/Executables/ghq/github.com/gohugoio/hugo/hugolib/shortcode_test.go:631
            s := buildSingleSite(t, deps.DepsCfg{WithTemplate: addTemplates, Fs: fs, Cfg: cfg}, BuildCfg{})
        
Source changed "content/en/section/section/section/bundle1/index.md": WRITE
Source changed "content/p1.md": WRITE
Source changed "content/mybundle/data.json": WRITE
Source changed "content/b1/index.md": WRITE
Markup type mmark is deprecated and will be removed in a future release. See https://gohugo.io//content-management/formats/#list-of-content-formats
Source changed "/tmp/hugo-modules-content-file241735885/README.md": WRITE
WARN 2021/06/11 18:28:41 Failed to read module config for "incompat3" in "/site/themes/incompat3/theme.toml": unmarshal failed: Near line 2 (last key parsed 'min_version'): Invalid float value: "0.55.0"
WARN 2021/06/11 18:28:41 Module "incompat1" is not compatible with this Hugo version; run "hugo mod graph" for more information.
WARN 2021/06/11 18:28:41 Module "incompat2" is not compatible with this Hugo version; run "hugo mod graph" for more information.
WARN 2021/06/11 18:28:42 Module "n1" is not compatible with this Hugo version; run "hugo mod graph" for more information.
FAIL
FAIL	github.com/gohugoio/hugo/hugolib	22.119s
ok  	github.com/gohugoio/hugo/hugolib/filesystems	(cached)
ok  	github.com/gohugoio/hugo/hugolib/paths	(cached)
ok  	github.com/gohugoio/hugo/identity	(cached)
ok  	github.com/gohugoio/hugo/langs	(cached)
ok  	github.com/gohugoio/hugo/langs/i18n	(cached)
ok  	github.com/gohugoio/hugo/lazy	(cached)
?   	github.com/gohugoio/hugo/livereload	[no test files]
ok  	github.com/gohugoio/hugo/markup	(cached)
ok  	github.com/gohugoio/hugo/markup/asciidocext	(cached)
?   	github.com/gohugoio/hugo/markup/asciidocext/asciidocext_config	[no test files]
ok  	github.com/gohugoio/hugo/markup/blackfriday	(cached)
?   	github.com/gohugoio/hugo/markup/blackfriday/blackfriday_config	[no test files]
?   	github.com/gohugoio/hugo/markup/converter	[no test files]
?   	github.com/gohugoio/hugo/markup/converter/hooks	[no test files]
ok  	github.com/gohugoio/hugo/markup/goldmark	(cached)
?   	github.com/gohugoio/hugo/markup/goldmark/goldmark_config	[no test files]
?   	github.com/gohugoio/hugo/markup/goldmark/internal/extensions/attributes	[no test files]
ok  	github.com/gohugoio/hugo/markup/highlight	(cached)
?   	github.com/gohugoio/hugo/markup/internal	[no test files]
ok  	github.com/gohugoio/hugo/markup/markup_config	(cached)
ok  	github.com/gohugoio/hugo/markup/mmark	(cached)
ok  	github.com/gohugoio/hugo/markup/org	(cached)
ok  	github.com/gohugoio/hugo/markup/pandoc	(cached)
ok  	github.com/gohugoio/hugo/markup/rst	(cached)
ok  	github.com/gohugoio/hugo/markup/tableofcontents	(cached)
--- FAIL: TestDefaultTypes (0.00s)
    mediaType_test.go:59: 
        error:
          values are not equal
        got:
          int(28)
        want:
          int(27)
        stack:
          /home/rkumar/Executables/ghq/github.com/gohugoio/hugo/media/mediaType_test.go:59
            c.Assert(len(DefaultTypes), qt.Equals, 27)
        
FAIL
FAIL	github.com/gohugoio/hugo/media	0.005s
ok  	github.com/gohugoio/hugo/metrics	(cached)
ok  	github.com/gohugoio/hugo/minifiers	(cached)
ok  	github.com/gohugoio/hugo/modules	(cached)
ok  	github.com/gohugoio/hugo/modules/npm	(cached)
ok  	github.com/gohugoio/hugo/navigation	(cached)
ok  	github.com/gohugoio/hugo/output	(cached)
ok  	github.com/gohugoio/hugo/parser	(cached)
ok  	github.com/gohugoio/hugo/parser/metadecoders	(cached)
ok  	github.com/gohugoio/hugo/parser/pageparser	(cached)
ok  	github.com/gohugoio/hugo/publisher	(cached)
ok  	github.com/gohugoio/hugo/related	(cached)
ok  	github.com/gohugoio/hugo/releaser	(cached)
ok  	github.com/gohugoio/hugo/resources	(cached)
ok  	github.com/gohugoio/hugo/resources/images	(cached)
ok  	github.com/gohugoio/hugo/resources/images/exif	(cached)
?   	github.com/gohugoio/hugo/resources/images/webp	[no test files]
ok  	github.com/gohugoio/hugo/resources/internal	(cached)
ok  	github.com/gohugoio/hugo/resources/jsconfig	(cached)
ok  	github.com/gohugoio/hugo/resources/page	(cached)
?   	github.com/gohugoio/hugo/resources/page/page_generate	[no test files]
ok  	github.com/gohugoio/hugo/resources/page/pagemeta	(cached)
ok  	github.com/gohugoio/hugo/resources/postpub	(cached)
?   	github.com/gohugoio/hugo/resources/resource	[no test files]
ok  	github.com/gohugoio/hugo/resources/resource_factories/bundler	(cached)
?   	github.com/gohugoio/hugo/resources/resource_factories/create	[no test files]
?   	github.com/gohugoio/hugo/resources/resource_transformers/babel	[no test files]
?   	github.com/gohugoio/hugo/resources/resource_transformers/htesting	[no test files]
ok  	github.com/gohugoio/hugo/resources/resource_transformers/integrity	(cached)
ok  	github.com/gohugoio/hugo/resources/resource_transformers/js	(cached)
ok  	github.com/gohugoio/hugo/resources/resource_transformers/minifier	(cached)
ok  	github.com/gohugoio/hugo/resources/resource_transformers/postcss	(cached)
?   	github.com/gohugoio/hugo/resources/resource_transformers/templates	[no test files]
?   	github.com/gohugoio/hugo/resources/resource_transformers/tocss/dartsass	[no test files]
ok  	github.com/gohugoio/hugo/resources/resource_transformers/tocss/scss	(cached)
?   	github.com/gohugoio/hugo/scripts/fork_go_templates	[no test files]
ok  	github.com/gohugoio/hugo/source	(cached)
ok  	github.com/gohugoio/hugo/tpl	(cached)
ok  	github.com/gohugoio/hugo/tpl/cast	(cached)
ok  	github.com/gohugoio/hugo/tpl/collections	(cached)
ok  	github.com/gohugoio/hugo/tpl/compare	(cached)
ok  	github.com/gohugoio/hugo/tpl/crypto	(cached)
ok  	github.com/gohugoio/hugo/tpl/data	(cached)
ok  	github.com/gohugoio/hugo/tpl/debug	(cached)
ok  	github.com/gohugoio/hugo/tpl/encoding	(cached)
ok  	github.com/gohugoio/hugo/tpl/fmt	(cached)
ok  	github.com/gohugoio/hugo/tpl/hugo	(cached)
ok  	github.com/gohugoio/hugo/tpl/images	(cached)
ok  	github.com/gohugoio/hugo/tpl/inflect	(cached)
ok  	github.com/gohugoio/hugo/tpl/internal	(cached)
?   	github.com/gohugoio/hugo/tpl/internal/go_templates/cfg	[no test files]
ok  	github.com/gohugoio/hugo/tpl/internal/go_templates/fmtsort	(cached)
ok  	github.com/gohugoio/hugo/tpl/internal/go_templates/htmltemplate	(cached)
?   	github.com/gohugoio/hugo/tpl/internal/go_templates/testenv	[no test files]
ok  	github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate	(cached)
ok  	github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/parse	(cached)
?   	github.com/gohugoio/hugo/tpl/internal/resourcehelpers	[no test files]
?   	github.com/gohugoio/hugo/tpl/js	[no test files]
ok  	github.com/gohugoio/hugo/tpl/lang	(cached)
ok  	github.com/gohugoio/hugo/tpl/math	(cached)
?   	github.com/gohugoio/hugo/tpl/openapi/openapi3	[no test files]
ok  	github.com/gohugoio/hugo/tpl/os	(cached)
ok  	github.com/gohugoio/hugo/tpl/partials	(cached)
ok  	github.com/gohugoio/hugo/tpl/path	(cached)
ok  	github.com/gohugoio/hugo/tpl/reflect	(cached)
?   	github.com/gohugoio/hugo/tpl/resources	[no test files]
ok  	github.com/gohugoio/hugo/tpl/safe	(cached)
ok  	github.com/gohugoio/hugo/tpl/site	(cached)
ok  	github.com/gohugoio/hugo/tpl/strings	(cached)
ok  	github.com/gohugoio/hugo/tpl/templates	(cached)
ok  	github.com/gohugoio/hugo/tpl/time	(cached)
ok  	github.com/gohugoio/hugo/tpl/tplimpl	(cached)
?   	github.com/gohugoio/hugo/tpl/tplimpl/embedded	[no test files]
?   	github.com/gohugoio/hugo/tpl/tplimpl/embedded/generate	[no test files]
ok  	github.com/gohugoio/hugo/tpl/transform	(cached)
ok  	github.com/gohugoio/hugo/tpl/urls	(cached)
ok  	github.com/gohugoio/hugo/transform	(cached)
ok  	github.com/gohugoio/hugo/transform/livereloadinject	(cached)
ok  	github.com/gohugoio/hugo/transform/metainject	(cached)
ok  	github.com/gohugoio/hugo/transform/urlreplacers	(cached)
?   	github.com/gohugoio/hugo/watcher	[no test files]
FAIL
```

Seirdy commented:
On Sat, Jun 12, 2021 at 07:14:35AM -0700, Bjørn Erik Pedersen wrote:
>OK, sorry for the confusion. The type should follow the naming standard 
>of the others (that is a discussion for another time :-); I meant the 
>`Name`,  which is now `Web App Manifest`.

Got it, fixed.

-- 
/Seirdy

bep commented on media/mediaType.go:
```diff
@@ -167,7 +167,7 @@ var (
 
 	JSONType           = newMediaType("application", "json", []string{"json"})
-	WebAppManifest = newMediaTypeWithMimeSuffix("application", "manifest", "json", []string{"webmanifest"})
+	WebAppManifestFormat = newMediaTypeWithMimeSuffix("application", "manifest", "json", []string{"webmanifest"})
```
This needs to be `WebAppManifestType` (see the other types).

Seirdy replied:
Done

bep commented:
You need to update the assertion in the test with the updated count.
